### ISSUE TITLE
关于学习MOE模块代码时，产生的一个问题

### ISSUE THREAD

**Issue by ZHJ19970917** [jingyaogong/minimind#340]
在MOEFeedForward类中的forward函数中：
```python
        if self.training:
            x = x.repeat_interleave(self.config.num_experts_per_tok, dim=0)
            y = torch.empty_like(x, dtype=torch.float16)
            for i, expert in enumerate(self.experts):
                y[flat_topk_idx == i] = expert(x[flat_topk_idx == i]).to(y.dtype)  # 确保类型一致
            y = (y.view(*topk_weight.shape, -1) * topk_weight.unsqueeze(-1)).sum(dim=1)
            y = y.view(*orig_shape)
```
这串代码最后将y的形状变为(*topk_weight.shape, -1)->(bsz, seq_len, top_k, dim)。在最后sum时，设定dim=1,那最后结果的形状是不是就变成了(bsz, top_k, dim)。最后期望的y的输出形状是不是应该是(bsz, seq_len, dim)，因为我看下一行是y = y.view(*orig_shape)，要将y变成orig_shape=x.shape->(bsz, seq_len, dim)

**Comment by jingyaogong** [comment]
```
x.repeat_interleave(self.config.num_experts_per_tok, dim=0)
=
torch.empty_like(x, dtype=torch.float16)
=
tensor(bsz * seq_len * num_experts_per_tok, dim)
```

```
(y.view(*topk_weight.shape, -1) * topk_weight.unsqueeze(-1))
=
tensor(bsz * seq_len, num_experts_per_tok, dim)
```

```
(y.view(*topk_weight.shape, -1) * topk_weight.unsqueeze(-1)).sum(dim=1)
=
tensor(bsz * seq_len, dim)
```

```
y.view(*orig_shape)
=
tensor(bsz, seq_len, dim)
```

**Comment by ZHJ19970917** [comment]
谢谢大佬!
